File: lims-frontend/src/api/testReports.ts
/**
 * Phiếu kết quả thử nghiệm (m46) — BM 7.8/01/RIBE.
 *
 * Module CHỨNG TỪ: tệp do Phòng nhận mẫu soạn ngoài hệ thống rồi tải lên. Ở đây theo
 * dõi bản nào đã ra khỏi Viện, lúc nào, ai chịu trách nhiệm, bản nào đã bị thay thế.
 *
 * Tải tệp về đi qua endpoint RIÊNG (không phải /attachments/{id}) vì mỗi lượt tải một
 * chứng từ đã trao khách đều phải để lại vết riêng — `revoke` dùng chính vết đó để
 * biết phải báo cho ai khi thu hồi.
 */
import { apiDelete, apiGet, apiPatch, apiPost, apiUpload } from '@/lib/api';
import type { DeliveryMethod, TestReport, TestReportFile } from '@/types';

export function listTestReports(intakeId: string) {
  return apiGet<TestReport[]>(`/intakes/${intakeId}/test-reports`);
}

export interface TestReportBody {
  /** Bỏ trống thì backend sinh theo mã phiếu nhận (26N323, rồi -2, -3 nếu trùng). */
  report_no?: string | null;
  title?: string | null;
  note?: string | null;
  issued_at?: string | null;
}

export function createTestReport(intakeId: string, body: TestReportBody) {
  return apiPost<TestReport>(`/intakes/${intakeId}/test-reports`, body);
}

export function updateTestReport(reportId: string, body: TestReportBody) {
  return apiPatch<TestReport>(`/test-reports/${reportId}`, body);
}

export function deleteTestReport(reportId: string) {
  return apiDelete(`/test-reports/${reportId}`);
}

/** Phát hành — từ đây tệp bị khoá, sửa nội dung phải tạo bản sửa đổi. */
export function issueTestReport(reportId: string, issuedAt?: string | null) {
  return apiPost<TestReport>(`/test-reports/${reportId}/issue`, {
    issued_at: issuedAt || null,
  });
}

export interface DeliverBody {
  delivery_method: DeliveryMethod;
  delivered_to?: string | null;
  delivered_at?: string | null;
  delivery_note?: string | null;
}

export function deliverTestReport(reportId: string, body: DeliverBody) {
  return apiPost<TestReport>(`/test-reports/${reportId}/deliver`, body);
}

export function revokeTestReport(reportId: string, reason: string) {
  return apiPost<TestReport>(`/test-reports/${reportId}/revoke`, { reason });
}

/** Bản sửa đổi = bản nháp mới version+1; bản cũ tự chuyển "Đã thu hồi". */
export function reviseTestReport(reportId: string, reason: string, reportNo?: string | null) {
  return apiPost<TestReport>(`/test-reports/${reportId}/revisions`, {
    reason,
    report_no: reportNo || null,
  });
}

export function uploadTestReportFile(reportId: string, file: File) {
  return apiUpload<TestReportFile>(`/test-reports/${reportId}/files`, file, 'file');
}

export function deleteTestReportFile(reportId: string, attachmentId: string) {
  return apiDelete(`/test-reports/${reportId}/files/${attachmentId}`);
}

/** Mở tệp trong tab mới bằng presigned URL. Lượt tải được ghi vào nhật ký (BR-13). */
export async function openTestReportFile(reportId: string, attachmentId: string) {
  const data = await apiGet<TestReportFile & { download_url: string }>(
    `/test-reports/${reportId}/files/${attachmentId}`,
  );
  if (data.download_url) window.open(data.download_url, '_blank', 'noopener');
}

File: lims-backend/app/routers/test_reports.py
"""Router PHIẾU KẾT QUẢ THỬ NGHIỆM (m46) — BM 7.8/01/RIBE.

Hai nhóm đường dẫn:
  · /intakes/{intake_id}/test-reports  — danh sách và lập mới cho một phiếu nhận mẫu
  · /test-reports/{report_id}/…        — vòng đời của một chứng từ cụ thể

Quyền do `test_report_service` kiểm (`test_report:read` / `test_report:manage`) chứ
không phải `require_permission` ở tầng router: cùng luật đó được `attachment_authz` gọi
lại cho đường generic `POST /attachments`, và đặt luật ở một chỗ là cách duy nhất để hai
đường không lệch nhau.

Mọi endpoint khai `response_model` — test kiến trúc chặn endpoint mới thiếu hợp đồng.
"""
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, File, Request, UploadFile, status
from sqlalchemy.orm import Session

from app.core.concurrency import upload_slot
from app.core.deps import CurrentUser, get_current_user
from app.core.request_meta import client_ip
from app.core.responses import ok
from app.db.database import get_db
from app.schemas.test_report import (
    CreateTestReportRequest,
    DeleteTestReportResponse,
    DeliverTestReportRequest,
    IssueTestReportRequest,
    ReviseTestReportRequest,
    RevokeTestReportRequest,
    TestReportDownloadResponse,
    TestReportFileResponse,
    TestReportListResponse,
    TestReportResponse,
    UpdateTestReportRequest,
)
from app.services import test_report_service as svc

router = APIRouter(tags=["m46-test-reports"])


def _cid(request: Request) -> Optional[str]:
    return getattr(request.state, "correlation_id", None)


# ═══════════════════════════ Theo phiếu nhận mẫu ═══════════════════════════


@router.get("/intakes/{intake_id}/test-reports", response_model=TestReportListResponse)
def list_test_reports(
    intake_id: uuid.UUID,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ok(svc.list_reports(db, user=user, intake_id=intake_id))


@router.post(
    "/intakes/{intake_id}/test-reports",
    response_model=TestReportResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_test_report(
    intake_id: uuid.UUID,
    body: CreateTestReportRequest,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Lập bản NHÁP. Tải tệp lên ở bước sau, rồi mới phát hành."""
    return ok(
        svc.create_report(
            db, user=user, intake_id=intake_id, fields=body.model_dump(),
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


# ═══════════════════════════ Một chứng từ ═══════════════════════════


@router.get("/test-reports/{report_id}", response_model=TestReportResponse)
def get_test_report(
    report_id: uuid.UUID,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ok(svc.get_report(db, user=user, report_id=report_id))


@router.patch("/test-reports/{report_id}", response_model=TestReportResponse)
def update_test_report(
    report_id: uuid.UUID,
    body: UpdateTestReportRequest,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ok(
        svc.update_report(
            db, user=user, report_id=report_id,
            changes=body.model_dump(exclude_unset=True),
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


@router.delete("/test-reports/{report_id}", response_model=DeleteTestReportResponse)
def delete_test_report(
    report_id: uuid.UUID,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Xoá MỀM, chỉ bản nháp. Đã phát hành thì dùng /revoke."""
    return ok(
        svc.delete_report(
            db, user=user, report_id=report_id,
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


@router.post("/test-reports/{report_id}/issue", response_model=TestReportResponse)
def issue_test_report(
    report_id: uuid.UUID,
    body: IssueTestReportRequest,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Phát hành — từ đây tệp bị khoá, muốn đổi nội dung phải tạo bản sửa đổi."""
    return ok(
        svc.issue_report(
            db, user=user, report_id=report_id, issued_at=body.issued_at,
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


@router.post("/test-reports/{report_id}/deliver", response_model=TestReportResponse)
def deliver_test_report(
    report_id: uuid.UUID,
    body: DeliverTestReportRequest,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Ghi nhận đã trao khách: ngày, hình thức, người nhận."""
    return ok(
        svc.deliver_report(
            db, user=user, report_id=report_id, fields=body.model_dump(),
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


@router.post("/test-reports/{report_id}/revoke", response_model=TestReportResponse)
def revoke_test_report(
    report_id: uuid.UUID,
    body: RevokeTestReportRequest,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Thu hồi bản đã phát hành. Báo cho mọi người đã từng tải tệp đó về."""
    return ok(
        svc.revoke_report(
            db, user=user, report_id=report_id, reason=body.reason,
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


@router.post(
    "/test-reports/{report_id}/revisions",
    response_model=TestReportResponse,
    status_code=status.HTTP_201_CREATED,
)
def revise_test_report(
    report_id: uuid.UUID,
    body: ReviseTestReportRequest,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Bản sửa đổi = bản NHÁP mới version+1; bản cũ tự chuyển 'revoked'."""
    return ok(
        svc.create_revision(
            db, user=user, report_id=report_id, reason=body.reason,
            report_no=body.report_no,
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


# ═══════════════════════════ Tệp ═══════════════════════════


@router.post(
    "/test-reports/{report_id}/files",
    response_model=TestReportFileResponse,
    status_code=status.HTTP_201_CREATED,
)
def upload_test_report_file(
    report_id: uuid.UUID,
    request: Request,
    file: UploadFile = File(...),
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    with upload_slot():
        content = file.file.read()
        return ok(
            svc.upload_file(
                db, user=user, report_id=report_id,
                file_name=file.filename or "file", content=content,
                mime=file.content_type,
                correlation_id=_cid(request), ip=client_ip(request),
            )
        )


@router.get(
    "/test-reports/{report_id}/files/{attachment_id}",
    response_model=TestReportDownloadResponse,
)
def download_test_report_file(
    report_id: uuid.UUID,
    attachment_id: uuid.UUID,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Cấp presigned URL và GHI VẾT ai đã lấy chứng từ ra khỏi hệ thống (BR-13)."""
    return ok(
        svc.download_file(
            db, user=user, report_id=report_id, attachment_id=attachment_id,
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )


@router.delete(
    "/test-reports/{report_id}/files/{attachment_id}",
    response_model=DeleteTestReportResponse,
)
def delete_test_report_file(
    report_id: uuid.UUID,
    attachment_id: uuid.UUID,
    request: Request,
    user: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return ok(
        svc.delete_file(
            db, user=user, report_id=report_id, attachment_id=attachment_id,
            correlation_id=_cid(request), ip=client_ip(request),
        )
    )
